clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,265 @@
+		}
+		if err = json.Unmarshal(dataByte, &tagContent); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+	tagContent.Tags[tag] = properties
+	err = saveTagData(req, localpart, roomID, accountDB, tagContent)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// send data to syncapi
+	if err := syncProducer.SendData(userID, roomID, "m.tag"); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+// Delete functionality works by obtaining the saved Tags, removing the intended tag from
+// the "map" and then saving the new "map" in the DB
+func DeleteTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot delete another user's tags"),
+		}
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// If there are no tags in the database, exit.
+	if len(data) == 0 {
+		//Synapse returns a 200 OK response on finding no Tags, same policy is followed here.
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: struct{}{},
+		}
+	}
+
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	var tagData []gomatrixserverlib.ClientEvent
+	tagContent := newTag()
+	err = json.Unmarshal(dataByte, &tagData)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	err = json.Unmarshal(tagData[0].Content, &tagContent)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// Check whether the Tag to be deleted exists
+	if _, ok := tagContent.Tags[tag]; ok {
+		delete(tagContent.Tags, tag)
+	} else {
+		//Synapse returns a 200 OK response on finding no Tags, same policy is followed here.
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: struct{}{},
+		}
+	}
+	err = saveTagData(req, localpart, roomID, accountDB, tagContent)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// send data to syncapi
+	if err := syncProducer.SendData(userID, roomID, "m.tag"); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// obtainSavedTags gets all the tags saved in the DB
+func obtainSavedTags(
+	req *http.Request,
+	userID string,
+	roomID string,
+	accountDB *accounts.Database,
+) (string, []gomatrixserverlib.ClientEvent, error) {
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return "", nil, err
+	}
+
+	data, err := accountDB.GetAccountDataByType(
```

Comment:
Since `GetAccountDataByType` will only return zero or one `ClientEvent` (due to the uniqueness of account data row associated with a given `(user, room, type)` tuple), you can consider retiring the array type here and just use a pointer instead.

Reply:
You mean I return `data[0]` ? (Of course after checking whether it is atleast of length 1)

Reply:
If you're going to use a pointer it'll be `&data[0]`, or `nil` if there's no data.

Reply:
Is this necessary? What problems does this cause? Because based on the length of this, different errors are raised in different functions. I'd probably let this be

Reply:
This is not a must and won't cause practical problems. It's just about different ways to do something, so yes you can leave it there.